### DISCUSSION TITLE
iOS Offer codes

### DISCUSSION THREAD

**Discussion by benglasgow** [hyochan/react-native-iap#2697] [General]
I read through the app store connect docs on how to set up an offer code, but I'm still confused on how to properly set up and use offer codes for a subscription.

I was able to get promo codes working and introductory codes working, but I'm looking to specifically set up offer codes.

My questions are:

1.) Do I use presentCodeRedemptionSheetIOS? If so, why do I keep getting an error that looks like this?

![image](https://github.com/dooboolab-community/react-native-iap/assets/160365847/fc701817-9cfe-47c3-9bc5-d77d0d1b8711)

2.) If I don't use a redemption sheet, what do I use to redeem offer codes?

3.) Is there some step that's not intuitive/documented that I might be missing when it comes to setting this up?
I have the offer code set up in app store connect, and I also have a custom code created for offer code.

I'd be happy to do some documenting to help people out with this in the future if that's helpful.

Any and all help is welcome.

Thank you!

**Comment by SufianBabri** [comment]
I've done some digging. This is unfortunately Apple's limitation that offer codes do not work in a sandboxed environment (i.e. TestFlight).

[Answer on Apple forums](https://developer.apple.com/forums/thread/688550?answerId=685921022#685921022):
> Offer Codes are not redeemable in Sandbox at this time but appreciate the feedback as improving the testing support in Sandbox is an ongoing effort. For the redemption test in production, I assume this redemption was using the in-app redemption sheet (presentCodeRedemptionSheet) initiated from your app. If the app is relaunched. Is the transaction now found? If so, please gather a sysdiagnose after reproducing and file a ticket at http://feedbackassistant.apple.com/  and please note here the ticket#.

[Answer on Apple forums](https://developer.apple.com/forums/thread/690560?answerId=709474022#709474022):
> You cannot test subscription offer codes in the Sandbox. 
> 
> When a customer successfully redeems an offer code from a live app in iOS 14 and later, the app receives a purchased transaction and its receipts contains the offer_code_ref_name field after validation. As such, redeeming an offer code creates a transaction in an app’s receipt in the same manner that a regular transaction occurs. However, this field is unsupported in receipts for apps tested in the Sandbox. However, you can simulate redemption that occurs outside of your app in the Sandbox by resubscribing using an Apple ID that subscribed in the past in the Manage Subscriptions page on an iOS 14 or later device.

I think the only way to test the offer code feature is to release the app to live which shows the "Redeem" button when the user has a specific identifier (email address or username). This way you can test the redemption flow without disrupting the flow of your live users. Not an ideal solution but there doesn't seem to be any other choice.

**Reply by benglasgow** [reply]
I'll give this a try today and report back. Thank you!

Y'know I thought it _could_ be something like that, but didn't want to test on the prod version of my app, lol.
